File: src/state/SessionStore.ts
import * as fs from "node:fs";
import * as path from "node:path";
import type { Message } from "src/types/message";
import { shouldRedact } from "./secretRedactor";

const REDACTION_PLACEHOLDER = "***";
const SCHEMA_VERSION = 1 as const;

export interface SessionRecord {
  schemaVersion: 1;
  id: string;
  createdAt: string;
  updatedAt: string;
  cwd: string;
  provider: string;
  model?: string;
  firstPrompt: string;
  stackNames: string[];
  messages: Message[]; // already redacted
}

export interface SessionIndexEntry {
  id: string;
  createdAt: string;
  updatedAt: string;
  firstPrompt: string;
  stackNames: string[];
}

/** Warn when resuming a session saved under a different working directory. */
export function sessionCwdMismatchWarning(record: SessionRecord, cwd: string): string | undefined {
  if (record.cwd === cwd) return undefined;
  return `Resuming session saved in ${record.cwd} (current directory: ${cwd}). Stack paths may differ.`;
}

/** Format index entries for /sessions output (newest-first). */
export function formatSessionsList(entries: SessionIndexEntry[]): string {
  if (entries.length === 0) return "No saved sessions.";
  return [
    "Saved sessions (newest first):",
    ...entries.map((entry, index) => {
      const stacks =
        entry.stackNames.length > 0 ? `  stacks: ${entry.stackNames.join(", ")}` : "";
      const prompt =
        entry.firstPrompt.length > 72
          ? `${entry.firstPrompt.slice(0, 69)}...`
          : entry.firstPrompt;
      return [
        `${index + 1}. ${entry.id}`,
        `   updated: ${entry.updatedAt}${stacks}`,
        `   prompt: ${prompt}`,
      ].join("\n");
    }),
    "",
    "Resume with /resume or /resume <id>",
  ].join("\n");
}

function warn(message: string): void {
  process.stderr.write(`[docker-agent] ${message}\n`);
}

function errorMessage(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}

/**
 * Recursively walk a `Message[]` and replace any string value whose key matches
 * SECRET_KEY_PATTERN with the redaction placeholder.
 *
 * The only place secrets appear in tool messages is in environment maps embedded
 * in tool-result content strings. We scan:
 *   - ToolResultMessage.content (string) — look for KEY=VALUE patterns
 *   - AssistantMessage content blocks (text, tool_use inputs as JSON)
 *
 * For simplicity and safety we do a string-level replacement: any JSON field
 * whose key matches SECRET_KEY_PATTERN has its value replaced with "***".
 */
function redactMessages(messages: Message[]): Message[] {
  return messages.map((msg) => {
    if (msg.role === "tool") {
      return { ...msg, content: redactString(msg.content) };
    }
    if (msg.role === "assistant") {
      return {
        ...msg,
        content: msg.content.map((block) => {
          if (block.type === "text") {
            return { ...block, text: redactString(block.text) };
          }
          if (block.type === "tool_use") {
            return { ...block, input: redactValue(block.input) };
          }
          return block;
        }),
      };
    }
    // UserMessage: content is a plain string; may contain injected env
    if (msg.role === "user") {
      return { ...msg, content: redactString(msg.content) };
    }
    return msg;
  });
}

/**
 * Redact a plain string — looks for `KEY=VALUE` or `"KEY": "VALUE"` patterns
 * where KEY matches the secret pattern.
 */
function redactString(input: string): string {
  // Replace JSON-style: "secretKey": "someValue"
  let result = input.replace(
    /"([^"]+)"(\s*:\s*)"([^"]*)"/g,
    (_match, key: string, colon: string, _value: string) => {
      if (shouldRedact(key)) {
        return `"${key}"${colon}"${REDACTION_PLACEHOLDER}"`;
      }
      return _match;
    },
  );
  // Replace shell-style: SECRET_KEY=someValue (unquoted or quoted)
  result = result.replace(
    /(\b\w+\b)(=)("[^"]*"|'[^']*'|[^\s,}]+)/g,
    (_match, key: string, eq: string, value: string) => {
      if (shouldRedact(key)) {
        return `${key}${eq}${REDACTION_PLACEHOLDER}`;
      }
      return _match;
    },
  );
  return result;
}

/**
 * Recursively redact an arbitrary JSON-serializable value (used for tool_use inputs).
 */
function redactValue(value: unknown): unknown {
  if (value === null || value === undefined) return value;
  if (typeof value === "string") return redactString(value);
  if (Array.isArray(value)) return value.map(redactValue);
  if (typeof value === "object") {
    const obj = value as Record<string, unknown>;
    const result: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(obj)) {
      result[k] = shouldRedact(k) ? REDACTION_PLACEHOLDER : redactValue(v);
    }
    return result;
  }
  return value;
}

export class SessionStore {
  private readonly sessionsDir: string;
  private readonly indexPath: string;

  constructor(root: string) {
    this.sessionsDir = path.join(root, "sessions");
    this.indexPath = path.join(this.sessionsDir, "index.json");
    fs.mkdirSync(this.sessionsDir, { recursive: true });
  }

  /**
   * Persist a session record atomically and upsert the index.
   * Messages are run through the secret redactor before writing.
   * If the write fails, logs a warning and preserves the prior file.
   */
  save(record: SessionRecord): void {
    const existing = this.read(record.id);
    const createdAt = existing?.createdAt ?? record.createdAt;
    const redacted: SessionRecord = {
      ...record,
      createdAt,
      messages: redactMessages(record.messages),
    };

    const filePath = path.join(this.sessionsDir, `${record.id}.json`);
    const tmpPath = `${filePath}.tmp`;

    try {
      fs.writeFileSync(tmpPath, JSON.stringify(redacted, null, 2), {
        encoding: "utf-8",
        mode: 0o644,
      });
      fs.renameSync(tmpPath, filePath);
    } catch (err) {
      // Clean up tmp file if it was created
      try {
        fs.unlinkSync(tmpPath);
      } catch {
        /* ignore cleanup errors */
      }
      warn(`SessionStore.save failed for session ${record.id}: ${errorMessage(err)}`);
      return; // preserve prior file, do not re-throw
    }

    // Upsert index
    this.upsertIndex({
      id: record.id,
      createdAt: record.createdAt,
      updatedAt: record.updatedAt,
      firstPrompt: record.firstPrompt,
      stackNames: record.stackNames,
    });
  }

  /**
   * Read a session record by id.
   * Returns null and warns on missing, corrupt, or wrong-schema files.
   * Never throws.
   */
  read(id: string): SessionRecord | null {
    const filePath = path.join(this.sessionsDir, `${id}.json`);
    try {
      if (!fs.existsSync(filePath)) {
        return null;
      }
      const raw = fs.readFileSync(filePath, "utf-8");
      const parsed: unknown = JSON.parse(raw);
      return this.validateRecord(parsed, filePath);
    } catch (err) {
      warn(`SessionStore.read failed for session ${id}: ${errorMessage(err)}`);
      return null;
    }
  }

  /**
   * Returns the most recently updated session record, or null if none exist.
   * Never throws.
   */
  latest(): SessionRecord | null {
    const entries = this.list();
    if (entries.length === 0) return null;
    // list() returns newest-first; first entry is the most recent
    const entry = entries[0];
    if (!entry) return null;
    return this.read(entry.id);
  }

  /**
   * Returns all index entries sorted newest-first by updatedAt.
   * Returns [] on missing or corrupt index.
   * Never throws.
   */
  list(): SessionIndexEntry[] {
    try {
      if (!fs.existsSync(this.indexPath)) {
        return [];
      }
      const raw = fs.readFileSync(this.indexPath, "utf-8");
      const parsed: unknown = JSON.parse(raw);
      if (!Array.isArray(parsed)) {
        warn("SessionStore.list: index.json is not an array, ignoring");
        return [];
      }
      const entries = parsed.filter(isValidIndexEntry);
      // Sort newest-first
      entries.sort((a, b) => new Date(b.updatedAt).getTime() - new Date(a.updatedAt).getTime());
      return entries;
    } catch (err) {
      warn(`SessionStore.list failed: ${errorMessage(err)}`);
      return [];
    }
  }

  // ---------------------------------------------------------------------------
  // Private helpers
  // ---------------------------------------------------------------------------

  private upsertIndex(entry: SessionIndexEntry): void {
    const tmpPath = `${this.indexPath}.tmp`;
    try {
      let entries: SessionIndexEntry[] = [];
      if (fs.existsSync(this.indexPath)) {
        try {
          const raw = fs.readFileSync(this.indexPath, "utf-8");
          const parsed: unknown = JSON.parse(raw);
          if (Array.isArray(parsed)) {
            entries = parsed.filter(isValidIndexEntry);
          }
        } catch {
          // Corrupt index — start fresh
          entries = [];
        }
      }

      // Exactly one entry per id — remove any prior entry for this id
      entries = entries.filter((e) => e.id !== entry.id);
      entries.push(entry);

      fs.writeFileSync(tmpPath, JSON.stringify(entries, null, 2), {
        encoding: "utf-8",
        mode: 0o644,
      });
      fs.renameSync(tmpPath, this.indexPath);
    } catch (err) {
      try {
        fs.unlinkSync(tmpPath);
      } catch {
        /* ignore */
      }
      warn(`SessionStore.upsertIndex failed: ${errorMessage(err)}`);
    }
  }

  private validateRecord(parsed: unknown, source: string): SessionRecord | null {
    if (typeof parsed !== "object" || parsed === null) {
      warn(`SessionStore: corrupt session file at ${source} (not an object)`);
      return null;
    }
    const obj = parsed as Record<string, unknown>;
    if (obj.schemaVersion !== SCHEMA_VERSION) {
      warn(
        `SessionStore: unrecognised schemaVersion ${String(obj.schemaVersion)} at ${source}; expected ${SCHEMA_VERSION}`,
      );
      return null;
    }
    // Basic structural validation — we trust the file if schemaVersion is correct
    if (
      typeof obj.id !== "string" ||
      typeof obj.createdAt !== "string" ||
      typeof obj.updatedAt !== "string" ||
      typeof obj.cwd !== "string" ||
      typeof obj.provider !== "string" ||
      typeof obj.firstPrompt !== "string" ||
      !Array.isArray(obj.stackNames) ||
      !Array.isArray(obj.messages)
    ) {
      warn(`SessionStore: corrupt session file at ${source} (missing required fields)`);
      return null;
    }
    return parsed as SessionRecord;
  }
}

function isValidIndexEntry(entry: unknown): entry is SessionIndexEntry {
  if (typeof entry !== "object" || entry === null) return false;
  const obj = entry as Record<string, unknown>;
  return (
    typeof obj.id === "string" &&
    typeof obj.createdAt === "string" &&
    typeof obj.updatedAt === "string" &&
    typeof obj.firstPrompt === "string" &&
    Array.isArray(obj.stackNames)
  );
}

File: src/docker_agent/state/session_store.py
"""JSON session persistence with secret redaction.

Parity: ``src/state/SessionStore.ts:1-345``.
"""

import contextlib
import json
import os
import shutil
import sys
from pathlib import Path
from typing import Any

from pydantic import TypeAdapter

from docker_agent.state.secret_redactor import (
    redact_text as _redact_string,
)
from docker_agent.state.secret_redactor import (
    redact_value_deep as _redact_value,
)
from docker_agent.types.message import (
    AssistantBlock,
    AssistantMessage,
    Message,
    ToolResultMessage,
    UserMessage,
)

SCHEMA_VERSION = 1

SessionRecord = dict[str, Any]
SessionIndexEntry = dict[str, Any]


def _error_message(err: object) -> str:
    return str(err) if isinstance(err, Exception) else str(err)


def _warn(message: str) -> None:
    sys.stderr.write(f"[docker-agent] {message}\n")


def redact_messages(messages: list[Message]) -> list[Message]:
    """Return a deep copy of ``messages`` with secret values scrubbed."""
    out: list[Message] = []
    for msg in messages:
        if isinstance(msg, UserMessage):
            out.append(UserMessage(content=_redact_string(msg.content)))
        elif isinstance(msg, ToolResultMessage):
            out.append(
                ToolResultMessage.model_validate(
                    {
                        "tool_use_id": msg.tool_use_id,
                        "content": _redact_string(msg.content),
                        "is_error": msg.is_error,
                    }
                )
            )
        elif isinstance(msg, AssistantMessage):
            blocks: list[Any] = []
            for block in msg.content:
                data = block.model_dump(by_alias=True)
                if data["type"] == "text":
                    blocks.append(
                        AssistantBlock.model_validate(
                            {"type": "text", "text": _redact_string(data["text"])}
                        )
                    )
                elif data["type"] == "tool_use":
                    blocks.append(
                        AssistantBlock.model_validate(
                            {
                                "type": "tool_use",
                                "id": data["id"],
                                "name": data["name"],
                                "input": _redact_value(data["input"]),
                            }
                        )
                    )
                else:
                    blocks.append(block)
            out.append(AssistantMessage(content=blocks))
        else:
            out.append(msg)
    return out


def session_cwd_mismatch_warning(record: SessionRecord, cwd: str) -> str | None:
    if record.get("cwd") == cwd:
        return None
    return (
        f"Resuming session saved in {record['cwd']} "
        f"(current directory: {cwd}). Stack paths may differ."
    )


def format_sessions_list(entries: list[SessionIndexEntry]) -> str:
    if not entries:
        return "No saved sessions."
    lines = ["Saved sessions (newest first):"]
    for index, entry in enumerate(entries):
        stacks = (
            f"  stacks: {', '.join(entry['stack_names'])}" if entry["stack_names"] else ""
        )
        prompt = entry["first_prompt"]
        if len(prompt) > 72:
            prompt = prompt[:69] + "..."
        lines.append(f"{index + 1}. {entry['id']}")
        lines.append(f"   updated: {entry['updated_at']}{stacks}")
        lines.append(f"   prompt: {prompt}")
    lines.append("")
    lines.append("Use /resume to pick a session")
    return "\n".join(lines)


class SessionStore:
    """Owns ``.docker-agent/sessions/<id>.json`` and ``index.json``."""

    def __init__(self, root: str | os.PathLike[str]) -> None:
        self._sessions_dir = Path(root) / "sessions"
        self._index_path = self._sessions_dir / "index.json"
        self._sessions_dir.mkdir(parents=True, exist_ok=True)

    def save(self, record: SessionRecord) -> None:
        existing = self.read(record["id"])
        created_at = existing["created_at"] if existing else record["created_at"]
        redacted_messages = redact_messages(
            TypeAdapter(list[Message]).validate_python(record["messages"])
        )
        redacted: SessionRecord = {
            **record,
            "created_at": created_at,
            "messages": [m.model_dump(by_alias=True) for m in redacted_messages],
        }
        file_path = self._sessions_dir / f"{record['id']}.json"
        tmp_path = Path(f"{file_path}.tmp")
        try:
            tmp_path.write_text(
                json.dumps(redacted, indent=2, ensure_ascii=False),
                encoding="utf-8",
            )
            os.chmod(tmp_path, 0o644)
            shutil.move(str(tmp_path), str(file_path))
        except Exception as err:  # noqa: BLE001
            with contextlib.suppress(FileNotFoundError):
                tmp_path.unlink()
            _warn(f"SessionStore.save failed for session {record['id']}: {_error_message(err)}")
            return
        self._upsert_index(
            {
                "id": record["id"],
                "created_at": record["created_at"],
                "updated_at": record["updated_at"],
                "first_prompt": record["first_prompt"],
                "stack_names": record["stack_names"],
            }
        )

    def read(self, session_id: str) -> SessionRecord | None:
        file_path = self._sessions_dir / f"{session_id}.json"
        if not file_path.exists():
            return None
        try:
            parsed = json.loads(file_path.read_text(encoding="utf-8"))
            return self._validate_record(parsed, str(file_path))
        except Exception as err:  # noqa: BLE001
            _warn(f"SessionStore.read failed for session {session_id}: {_error_message(err)}")
            return None

    def latest(self) -> SessionRecord | None:
        entries = self.list()
        if not entries:
            return None
        return self.read(entries[0]["id"])

    def list(self) -> list[SessionIndexEntry]:
        if not self._index_path.exists():
            return []
        try:
            parsed = json.loads(self._index_path.read_text(encoding="utf-8"))
            if not isinstance(parsed, list):
                _warn("SessionStore.list: index.json is not an array, ignoring")
                return []
            entries = [e for e in parsed if self._is_valid_index_entry(e)]
            entries.sort(
                key=lambda e: e["updated_at"], reverse=True
            )
            return entries
        except Exception as err:  # noqa: BLE001
            _warn(f"SessionStore.list failed: {_error_message(err)}")
            return []

    def _validate_record(self, obj: object, source: str) -> SessionRecord | None:
        if not isinstance(obj, dict):
            _warn(f"SessionStore: corrupt session file at {source} (not an object)")
            return None
        if obj.get("schema_version") != SCHEMA_VERSION:
            _warn(
                "SessionStore: unrecognised schemaVersion "
                f"{obj.get('schema_version')} at {source}; expected {SCHEMA_VERSION}"
            )
            return None
        required = [
            "id",
            "created_at",
            "updated_at",
            "cwd",
            "provider",
            "first_prompt",
            "stack_names",
            "messages",
        ]
        if not all(k in obj and isinstance(obj[k], type_for_key(k)) for k in required):
            _warn(f"SessionStore: corrupt session file at {source} (missing required fields)")
            return None
        return obj

    def _is_valid_index_entry(self, obj: object) -> bool:
        if not isinstance(obj, dict):
            return False
        return (
            isinstance(obj.get("id"), str)
            and isinstance(obj.get("created_at"), str)
            and isinstance(obj.get("updated_at"), str)
            and isinstance(obj.get("first_prompt"), str)
            and isinstance(obj.get("stack_names"), list)
        )

    def _upsert_index(self, entry: SessionIndexEntry) -> None:
        entries: list[SessionIndexEntry] = []
        if self._index_path.exists():
            try:
                parsed = json.loads(self._index_path.read_text(encoding="utf-8"))
                if isinstance(parsed, list):
                    entries = parsed
            except Exception:  # noqa: BLE001
                pass
        entries = [e for e in entries if e.get("id") != entry["id"]]
        entries.append(entry)
        tmp_path = Path(f"{self._index_path}.tmp")
        try:
            tmp_path.write_text(
                json.dumps(entries, indent=2, ensure_ascii=False), encoding="utf-8"
            )
            os.chmod(tmp_path, 0o644)
            shutil.move(str(tmp_path), str(self._index_path))
        except Exception as err:  # noqa: BLE001
            with contextlib.suppress(FileNotFoundError):
                tmp_path.unlink()
            _warn(f"SessionStore.upsertIndex failed: {_error_message(err)}")


def type_for_key(key: str) -> type:
    mapping: dict[str, type] = {
        "id": str,
        "created_at": str,
        "updated_at": str,
        "cwd": str,
        "provider": str,
        "first_prompt": str,
        "stack_names": list,
        "messages": list,
    }
    return mapping.get(key, object)


__all__ = [
    "SessionIndexEntry",
    "SessionRecord",
    "SessionStore",
    "format_sessions_list",
    "redact_messages",
    "session_cwd_mismatch_warning",
]